pkg/executor/adapter.go

```diff
@@ -664,6 +664,14 @@ func (a *ExecStmt) handleStmtForeignKeyTrigger(ctx context.Context, e exec.Execu
 		// Since `UnionScanExec` use `SnapshotIter` and `SnapshotGetter` to read txn mem-buffer, if we don't do `StmtCommit`,
 		// then the fk cascade executor can't read the mem-buffer changed by the ExecStmt.
 		a.Ctx.StmtCommit(ctx)
+		// `StmtCommit` will change the snapshot of the MemBuffer, so we need to get a new one.
```

Comment:
Should we wrap the `MembufferSnapshot` refreshing into the `StmtCommit` and `StmtRollback`, or wrap it into a function?

Reply:
`GetSnapshot` starts a new statement, and `StmtCommit` and `StmtRollback` ends up the lifetime of a statement. It's strange to auto start a new statement when ending previous one.

The use of `StmtCommit` here is an exception. If we redefine `GetSnapshot` to return the latest snapshot instead of `stages[0]`, the ideal implementation here can be:

```go
func (a *ExecStmt) handleStmtForeignKeyTrigger(...) {
	if stmtCtx.ForeignKeyTriggerCtx.HasFKCascades {
		a.Ctx.GetSessionVars().StmtCtx.MemBufferSnapshot = txn.GetMemBuffer().GetSnapshot()
	}
	...
}
```

Then if there are any pessimistic conflicts in foreign key handling, we can rollback the statement and retry it.